aptos-move/framework/aptos-framework/sources/storage_gas.move

```diff
@@ -209,32 +233,45 @@ module aptos_framework::storage_gas {
         // Check the corner case that current_usage_bps drops before the first point.
         let (left, right) = if (num_points == 0) {
             (&Point {x: 0, y: 0}, &Point {x: BASIS_POINT_DENOMINATION, y: BASIS_POINT_DENOMINATION})
+        } else if (current_usage_bps < vector::borrow(points, 0).x) {
+            (&Point { x: 0, y: 0 }, vector::borrow(points, 0))
+        } else if (vector::borrow(points, num_points - 1).x <= current_usage_bps) {
+            (vector::borrow(points, num_points-1), &Point { x: BASIS_POINT_DENOMINATION, y: BASIS_POINT_DENOMINATION })
         } else {
-            let (i, j) = (0, num_points - 1);
-            while (i < j) {
+            let (i, j) = (0, num_points - 2);
+            while ({
+                spec {
+                    invariant i <= j;
+                    invariant j < num_points;
+                    invariant points[i].x <= current_usage_bps;
+                    invariant current_usage_bps < points[j+1].x;
+                };
+                i < j
+            }) {
                 let mid = j - (j - i) / 2;
                 if (current_usage_bps < vector::borrow(points, mid).x) {
+                    spec {
+                        // j is strictly decreasing.
+                        assert mid-1 < j;
+                    };
                     j = mid - 1;
                 } else {
+                    spec {
+                        // i is strictly increasing.
+                        assert i < mid;
+                    };
                     i = mid;
                 };
             };
-            if (current_usage_bps < vector::borrow(points, 0).x) {
-                (&Point {x: 0, y: 0}, vector::borrow(points, 0))
-            } else {
-                // Check the corner case that current_usage_bps drops after the last point.
-                (
-                    vector::borrow(points, i),
-                    if (i == num_points - 1) {
-                        &Point { x: BASIS_POINT_DENOMINATION, y: BASIS_POINT_DENOMINATION }
-                    } else {
-                        vector::borrow(points, i + 1)
-                    }
-                )
-            }
+            (vector::borrow(points, i), vector::borrow(points, i + 1))
         };
+        let y_interpolated = interpolate(left.x, right.x, left.y, right.y, current_usage_bps);
+        interpolate(0, BASIS_POINT_DENOMINATION, curve.min_gas, curve.max_gas, y_interpolated)
+    }
 
-        curve.min_gas + (curve.max_gas - curve.min_gas) * (left.y + (current_usage_bps - left.x) * (right.y - left.y) / (right.x - left.x)) / BASIS_POINT_DENOMINATION
+    fun interpolate(x0: u64, x1: u64, y0: u64, y1: u64, x: u64): u64 {
+        let y = y0 + (x - x0) * (y1 - y0) / (x1 - x0);
```

Comment:
why cannot we omit `y`?

Reply:
We can. I omitted `y` and added a comment instead.